File: 0678_validParenthesisString/main.py
class Solution:
    def checkValidString(self, s: str) -> bool:
        minCount, maxCount = 0, 0

        for c in s:
            if c == '(':
                minCount += 1
                maxCount += 1
            if c == ')':
                minCount -= 1
                maxCount -= 1
            if c == '*':
                minCount -= 1
                maxCount += 1
            
            if maxCount < 0:
                return False
            if minCount < 0: # required s = (*)(
                minCount = 0
        return minCount == 0

File: 0678_validParenthesisString/main.cpp
#include <string>
#include <stack>

class Solution {

public:
    bool checkValidString(std::string s) {
        int min = 0;
        int max = 0;
        for(char &c : s) {
            if(c == '(') {
                ++min;
                ++max;
            } else if(c == ')') {
                --min;
                --max;
            } else if(c == '*') {
                --min;
                ++max;
            }

            if(max < 0)
                return false;
            if(min < 0)
                min = 0;
        }
        return min == 0;
    }
};


int main() {
    std::string test("(*))");
    Solution sol;
    sol.checkValidString(test);
    return 0;
}